src/reference/asciidoc/testing.adoc

```diff
@@ -2,12 +2,278 @@
+    assertThat(result, is(notNullValue()));
+    assertThat(result, hasPayload(is(instanceOf(byte[].class))));
+    assertThat(result, hasPayload(SAMPLE_CONTENT.getBytes(DEFAULT_ENCODING)));
+}
+----
+
+The `MockitoMessageMatchers` factory can be used for mocks stubbing and verifications:
+
+[source,java]
+----
+static final Date SOME_PAYLOAD = new Date();
+
+static final String SOME_HEADER_VALUE = "bar";
+
+static final String SOME_HEADER_KEY = "test.foo";
+...
+Message<?> message = MessageBuilder.withPayload(SOME_PAYLOAD)
+                .setHeader(SOME_HEADER_KEY, SOME_HEADER_VALUE)
+                .build();
+MessageHandler handler = mock(MessageHandler.class);
+handler.handleMessage(message);
+verify(handler).handleMessage(messageWithPayload(SOME_PAYLOAD));
+verify(handler).handleMessage(messageWithPayload(is(instanceOf(Date.class))));
+...
+MessageChannel channel = mock(MessageChannel.class);
+when(channel.send(messageWithHeaderEntry(SOME_HEADER_KEY, is(instanceOf(Short.class)))))
+        .thenReturn(true);
+assertThat(channel.send(message), is(false));
+----
+
+More utils can be added or migrated eventually.
+For example `RemoteFileTestSupport` implementations for the (S)FTP tests can be moved from the `test` directory of the particular modules to this publicly available `spring-integration-test-support` artifact.
+
+[[test-context]]
+=== Spring Integration and test context
+
+Typically tests for Spring applications use the Spring Test Framework and since Spring Integration is fully based on the Spring Framework foundation, everything what we can do with the Spring Test Framework is applied as well when we come to the task for testing integration flows.
+The `org.springframework.integration.test.context` package provides some components for enhancing test context for integration needs.
+First of all we should configure our test class with particular `@SpringIntegrationTest` annotation to enable Spring Integration Test Framework:
+
+[source,java]
+----
+@RunWith(SpringRunner.class)
+@SpringIntegrationTest(noAutoStartup = {"inboundChannelAdapter", "*Source*"})
+public class MyIntegrationTests {
+
+    @Autowired
+    private MockIntegrationContext mockIntegrationContext;
+
+}
+----
+
+The `@SpringIntegrationTest` populates `MockIntegrationContext` bean which can be autowired to the test class for further usage.
+With the provided `noAutoStartup` option, the Spring Integration Test Framework switches off endpoints, which are matched to the provided patterns, from the auto-startup during application context refreshing.
+This is useful, when we would like do not have real connections to the target systems from Inbound Channel Adapters, for example AMQP Inbound Gateway, JDBC Polling Channel Adapter, WebSocket Message Producer in client mode etc.
+
+The `MockIntegrationContext` is aimed to be used in the target test-cases for modifications beans in the real application context, for example those stopped endpoints can be replaced with mocks:
```

Comment:
> those stopped endpoints

What if those endpoints were not stopped by `noAutoStartup`? (because for example the user forgot to include them into the annotation parameter or mistaken while doing it) AFAIU, the endpoint will be relocated into `MockIntegrationContext#beans` but will continue to act as before.
Maybe we should explicitly stop the endpoint being replaced? Or at least warn the users here in documentation that replacing an endpoint doesn't stop it?